File: sicora-be-python/evalinservice/app/domain/value_objects/evaluation_status.py
"""Evaluation status value object."""

from enum import Enum


class EvaluationStatus(Enum):
    """
    Value object que define los estados posibles de una evaluación.
    """
    DRAFT = "draft"  # Borrador, en proceso de creación
    SUBMITTED = "submitted"  # Enviada para procesamiento
    PROCESSED = "processed"  # Procesada y agregada a estadísticas
    ARCHIVED = "archived"  # Archivada
    
    @classmethod
    def get_display_name(cls, status: "EvaluationStatus") -> str:
        """Retorna el nombre para mostrar del estado."""
        display_names = {
            cls.DRAFT: "Borrador",
            cls.SUBMITTED: "Enviada",
            cls.PROCESSED: "Procesada",
            cls.ARCHIVED: "Archivada"
        }
        return display_names.get(status, status.value)
    
    @classmethod
    def get_description(cls, status: "EvaluationStatus") -> str:
        """Retorna una descripción del estado."""
        descriptions = {
            cls.DRAFT: "La evaluación está en proceso de creación",
            cls.SUBMITTED: "La evaluación ha sido enviada para procesamiento",
            cls.PROCESSED: "La evaluación ha sido procesada y agregada a las estadísticas",
            cls.ARCHIVED: "La evaluación ha sido archivada"
        }
        return descriptions.get(status, "Estado desconocido")
    
    @classmethod
    def can_be_modified(cls, status: "EvaluationStatus") -> bool:
        """Verifica si la evaluación puede ser modificada en este estado."""
        return status == cls.DRAFT
    
    @classmethod
    def is_complete(cls, status: "EvaluationStatus") -> bool:
        """Verifica si la evaluación está completa."""
        return status in [cls.SUBMITTED, cls.PROCESSED, cls.ARCHIVED]
    
    @classmethod
    def is_counted_in_statistics(cls, status: "EvaluationStatus") -> bool:
        """Verifica si la evaluación debe ser contada en estadísticas."""
        return status in [cls.SUBMITTED, cls.PROCESSED]

File: sicora-be-go/evalinservice/internal/domain/valueobjects/evaluation_status.go
package valueobjects

// EvaluationStatus representa el estado de una evaluación
type EvaluationStatus string

const (
	EvaluationStatusDraft     EvaluationStatus = "DRAFT"     // Borrador, en progreso
	EvaluationStatusSubmitted EvaluationStatus = "SUBMITTED" // Enviada, completa
	EvaluationStatusValidated EvaluationStatus = "VALIDATED" // Validada por administrador
	EvaluationStatusCompleted EvaluationStatus = "COMPLETED" // Completada (alias para submitted)
	EvaluationStatusPending   EvaluationStatus = "PENDING"   // Pendiente
)

// IsValid verifica si el estado de la evaluación es válido
func (es EvaluationStatus) IsValid() bool {
	switch es {
	case EvaluationStatusDraft, EvaluationStatusSubmitted, EvaluationStatusValidated, EvaluationStatusCompleted, EvaluationStatusPending:
		return true
	default:
		return false
	}
}

// String retorna la representación en string del estado
func (es EvaluationStatus) String() string {
	return string(es)
}

// CanBeModified indica si la evaluación puede ser modificada
func (es EvaluationStatus) CanBeModified() bool {
	return es == EvaluationStatusDraft
}

// CanBeSubmitted indica si la evaluación puede ser enviada
func (es EvaluationStatus) CanBeSubmitted() bool {
	return es == EvaluationStatusDraft
}

// CanBeValidated indica si la evaluación puede ser validada
func (es EvaluationStatus) CanBeValidated() bool {
	return es == EvaluationStatusSubmitted
}

// IsComplete indica si la evaluación está completa
func (es EvaluationStatus) IsComplete() bool {
	return es == EvaluationStatusSubmitted || es == EvaluationStatusValidated
}
